### ISSUE TITLE
Should JSON output truncate the propertyUrl in cases where propertyUrl is not specifically defined?

### ISSUE THREAD

**Issue by 6a6d74** [w3c/csvw#298]
In the csv2json document, I am proposing the following text for specifying the `name` in the name-value pair describing the value of the cells ...

> For each cell [...] insert a name-value pair:
> 
> Name _N_ is determined from the propertyUrl property.
> 
> If the value of propertyUrl for the cell has not been explicitly specified in the metadata description (e.g. the value of propertyUrl is a concatenation of the CSV+ file URL, "#" (U+0023) and the value of the <a href="http://www.w3.org/TR/tabular-metadata/#column-name"><code>name</code></a> property for the column associated with the cell), then name _N_ SHALL take the value of the only the fragment identifier of propertyUrl.
> 
> Else, <a>name</a> <var>N</var> SHALL take the entire value of <a>propertyUrl</a>.

This means that you get "nice" looking JSON like:

```
[{ 
    "countryCode": "AD", 
    "latitude": "42.546245", 
    "longitude": "1.601554", 
    "name": "Andorra" 
}, { 
    "countryCode": "AE", 
    "latitude": "23.424076", 
    "longitude": "53.847818", 
    "name": "United Arab Emirates" 
}, { 
    "countryCode": "AF", 
    "latitude": "33.93911", 
    "longitude": "67.709953", 
    "name": "Afghanistan" 
}]
```

instead of having embedded URLs everywhere, like:

```
[{ 
    "http://example.org/countries.csv#countryCode": "AD", 
    "http://example.org/countries.csv#latitude": "42.546245", 
    "http://example.org/countries.csv#longitude": "1.601554", 
    "http://example.org/countries.csv#name": "Andorra" 
}, { 
    "http://example.org/countries.csv#countryCode": "AE", 
    "http://example.org/countries.csv#latitude": "23.424076", 
    "http://example.org/countries.csv#longitude": "53.847818", 
    "http://example.org/countries.csv#name": "United Arab Emirates" 
}, { 
    "http://example.org/countries.csv#countryCode": "AF", 
    "http://example.org/countries.csv#latitude": "33.93911", 
    "http://example.org/countries.csv#longitude": "67.709953", 
    "http://example.org/countries.csv#name": "Afghanistan" 
}]
```

Comments please ...

**Comment by iherman** [comment]
+1

Ivan

> On 05 Mar 2015, at 03:22 , Jeremy Tandy notifications@github.com wrote:
> 
> In the csv2json document, I am proposing the following text for specifying the name in the name-value pair describing the value of the cells ...
> 
> For each cell [...] insert a name-value pair:
> 
> Name N is determined from the propertyUrl property.
> If the value of propertyUrl for the cell has not been explicitly specified in the metadata description (e.g. the value of propertyUrl is a concatenation of the CSV+ file URL, "#" (U+0023) and the value of the name property for the column associated with the cell), then name N SHALL take the value of the only the fragment identifier of propertyUrl.
> Else, name N SHALL take the entire value of propertyUrl.
> 
> This means that you get "nice" looking JSON like:
> 
> [{
>     "countryCode": "AD",
>     "latitude": "42.546245",
>     "longitude": "1.601554",
>     "name": "Andorra"
> }, {
>     "countryCode": "AE",
>     "latitude": "23.424076",
>     "longitude": "53.847818",
>     "name": "United Arab Emirates"
> }, {
>     "countryCode": "AF",
>     "latitude": "33.93911",
>     "longitude": "67.709953",
>     "name": "Afghanistan"
> }]
> 
> instead of having embedded URLs everywhere, like:
> 
> [{
>     "http://example.org/countries.csv#countryCode": "AD",
>     "http://example.org/countries.csv#latitude": "42.546245",
>     "http://example.org/countries.csv#longitude": "1.601554",
>     "http://example.org/countries.csv#name": "Andorra"
> }, {
>     "http://example.org/countries.csv#countryCode": "AE",
>     "http://example.org/countries.csv#latitude": "23.424076",
>     "http://example.org/countries.csv#longitude": "53.847818",
>     "http://example.org/countries.csv#name": "United Arab Emirates"
> }, {
>     "http://example.org/countries.csv#countryCode": "AF",
>     "http://example.org/countries.csv#latitude": "33.93911",
>     "http://example.org/countries.csv#longitude": "67.709953",
>     "http://example.org/countries.csv#name": "Afghanistan"
> }]
> 
> Comments please ...
> 
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by JeniT** [comment]
This is fine, but it makes me wonder whether we're setting `propertyUrl` in the right place. Perhaps we should say that `propertyUrl` is `null` by default. Then we could leave the RDF conversion to construct a default property URL based on the name of the column, and the JSON conversion wouldn't have this amount of work to do.

**Comment by 6a6d74** [comment]
@JeniT - this is a good option. Of course, then the RDF conversion needs to _construct_ the `propertyUrl` ... it's 50-50% (perhaps 45-55 in favour of your suggestion). If I get three +1s to your suggestion, I will make the change.

**Comment by iherman** [comment]
I am not sure... At present, the conversion documents do not even refer to the template mechanism, how it is expanded, etc. Of course, the RDF conversion can refer back to the metadata document but it looks more complicated to me...

So my answer is '0' :-) meaning that maybe @JeniT and you can try out what it would mean, editorially, and then choose among the less complex approach

Ivan

> On 07 Mar 2015, at 04:01 , Jeremy Tandy notifications@github.com wrote:
> 
> @JeniT - this is a good option. Of course, then the RDF conversion needs to construct the propertyUrl ... it's 50-50% (perhaps 45-55 in favour of your suggestion). If I get three +1s to your suggestion, I will make the change.
> 
> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by JeniT** [comment]
The conversion documents wouldn't have to refer to URI templates: you can construct a URL by string concatenation (`url` of the table + `#` + `name` of the column).

**Comment by gkellogg** [comment]
I think leaving propertyUrl in the metadata is the right place. JSON can say that all propertlyUrls are shorted by removing URL + "#" if it is a prefix.

**Comment by JeniT** [comment]
Note: the [model document](http://w3c.github.io/csvw/syntax/#converting-tables) currently says that conversion specifications MUST use the `propertyUrl`. The [metadata document](http://w3c.github.io/csvw/metadata/#column-name) says that conversion specifications MUST use the `name`.

Should we soften the language here to say that conversion specifications MUST use _either_ the `name` or the `propertyUrl`. That would leave it open for the JSON output to just use the `name` without having to do URL hacking (potentially, maybe).

Either way, the language in those two specifications needs to be brought into line.

**Comment by iherman** [comment]
> On 11 Mar 2015, at 10:25 , Jeni Tennison notifications@github.com wrote:
> 
> Note: the model document currently says that conversion specifications MUST use the propertyUrl. The metadata document says that conversion specifications MUST use the name.

Hm. I am not sure I even understand that where this discrepancy comes from :-( My understanding is that
- if there is a propertyUrl, that takes the precedence and MUST be used
- otherwise, name MUST be used

I think the conversion specification language is based on the fact that, in the final metadata, this distinction disappears, ie, the propertyUrl value is always set in the metadata (or is returned by the value?), that may be the reason of this issue. (Sorry, I am at a conference right now, I cannot really go into the details.) Maybe it would make the conversion specification simpler (at least for JSON) if both those values would be kept down to the conversion level (in which case the slightly convoluted approach taken in the current JSON document could be simplified).

> Should we soften the language here to say that conversion specifications MUST use either the name or the propertyUrl. That would leave it open for the JSON output to just use the name without having to do URL hacking (potentially, maybe).

I think leaving it open for JSON will become messy and would hurt interoperability.

> Either way, the language in those two specifications needs to be brought into line.

+1

Ivan

> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by gkellogg** [comment]
RESOLUTION: propertyUrl can be null; the RDF conversion will construct a default based on the table URL & column name; JSON can use the name of the column if there’s no propertyUrl; we say what conversion specifications must say _only_ in the model document, and say they must use propertyUrl if provided, otherwise name

**Comment by 6a6d74** [comment]
@iherman - you said: 

> I think the conversion specification language is based on the fact that, in the final metadata, this distinction disappears, ie, the propertyUrl value is always set in the metadata [...]

This was my concern ... by the time the conversions occur `propertyUrl` is _always_ set. 

See [definition from the metadata document for property `propertyUrl`](http://w3c.github.io/csvw/metadata/#cell-propertyUrl):

> The default value of `propertyUrl` is `"{#_name}"`.

**Comment by 6a6d74** [comment]
Given the resolution, I will update csv2rdf and csv2json ...

**Comment by 6a6d74** [comment]
@gkellogg - I've updated (& merged changes; #346) in csv2json and csv2rdf ... now just awaiting the changes in the metadata and model documents. Over to you ...
